odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,731 @@
+
+    The squared L2-norm, ``||x||_2^2``,  is defined as the integral/sum of
+    ``x^2``.
+
+    References
+    ----------
+    Wikipedia article on `Norms
+        <https://en.wikipedia.org/wiki/Norm_(mathematics)>`_."""
+
+    def __init__(self, domain):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `LinearSpace`
+            Domain of the functional.
+        """
+        super().__init__(domain=domain, linear=False, grad_lipschitz=2)
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Parameters
+        ----------
+        x : `LinearSpaceVector`
+            Element in the domain of the functional.
+
+        Returns
+        -------
+        out : `field` element
+            Result of the evaluation.
+        """
+        return x.inner(x)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        functional = self
+
+        class L2SquareGradient(Operator):
+
+            """Gradient operator of this functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=True)
+
+            def _call(self, x):
+                """Applies the gradient operator to the given point.
+
+                Parameters
+                ----------
+                x : `LinearSpaceVector`
+                    Element in the domain of the functional to which the
+                    gradient operator is applied.
+
+                Returns
+                -------
+                out : `LinearSpaceVector`
+                    Result of the evaluation of the gradient operator. An
+                    element in the domain of the functional.
+                """
+                return 2.0 * x
+
+        return L2SquareGradient()
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Step size-like parameter
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+        return proximal_l2_squared(space=self.domain)(sigma)
+
+    @property
+    def conjugate_functional(self):
+        """The convex conjugate functional of the squared L2-norm.
+
+        Notes
+        -----
+        The conjugate functional of :math:`\|x\|_2^2` is
+        :math:`\\frac{1}{4}\|x\|_2^2`
+        """
+        return 1.0 / 4.0 * L2NormSquare(domain=self.domain)
```

Comment:
While python is associative from the left, this is non-obvious to even more experianced users (see: me). Hence I'd either write `0.25 * ` or `(1 / 4) *`